src/R/Editor/Impl/Completion/Providers/FilesCompletionProvider.cs

```diff
@@ -41,35 +49,42 @@ internal sealed class FilesCompletionProvider : IRCompletionListProvider {
             _glyphService = glyphService;
             _forceR = forceR;
 
-            _directory = ExtractDirectory(directoryCandidate);
-
-            if (_directory.Length == 0 || _directory.StartsWithOrdinal("~\\")) {
-                _directory = _directory.Length > 1 ? _directory.Substring(2) : _directory;
-                _userDirectoryFetchingTask = _workflow.RSession.GetRUserDirectoryAsync();
+            _enteredDirectory = ExtractDirectory(directoryCandidate);
+            if (_enteredDirectory.Length == 0 || _enteredDirectory.StartsWithOrdinal(".")) {
+                _mode = Mode.WorkingDirectory;
+                _task = Task.Run(async () => _rootDirectory = await _workflow.RSession.GetWorkingDirectoryAsync());
+            } else if (_enteredDirectory.StartsWithOrdinal("~\\")) {
+                _mode = Mode.UserDirectory;
+                _task = Task.Run(async () => _rootDirectory = await _workflow.RSession.GetRUserDirectoryAsync());
             }
+            _task?.DoNotWait();
         }
 
         #region IRCompletionListProvider
         public bool AllowSorting { get; } = false;
 
         public IReadOnlyCollection<RCompletion> GetEntries(RCompletionContext context) {
-            List<RCompletion> completions = new List<RCompletion>();
-            string directory = null;
-            string userDirectory = null;
+            var completions = new List<RCompletion>();
+            string directory = _enteredDirectory;
 
-            if (_userDirectoryFetchingTask != null) {
-                userDirectory = _userDirectoryFetchingTask.WaitTimeout(500);
-                userDirectory = userDirectory ?? _cachedUserDirectory;
-            }
+            try {
+                // If we are running async directory fetching, wait a bit
+                _task?.Wait(500);
+            } catch (OperationCanceledException) { }
 
             try {
-                if (!string.IsNullOrEmpty(userDirectory)) {
-                    _cachedUserDirectory = userDirectory;
-                    directory = Path.Combine(userDirectory, _directory);
-                } else {
-                    directory = _directory;
+                // If directory is set, then the async task did complete
+                if (!string.IsNullOrEmpty(_rootDirectory)) {
+                    if (_mode == Mode.WorkingDirectory) {
+                        directory = Path.Combine(_rootDirectory, _enteredDirectory);
+                    } else if (_mode == Mode.UserDirectory) {
+                        var subDirectory = _enteredDirectory.Length > 1 ? _enteredDirectory.Substring(2) : _enteredDirectory;
+                        directory = Path.Combine(_rootDirectory, subDirectory);
+                    }
                 }
+            } catch (ArgumentException) { }
```

Comment:
It comes from the user so possible. Also, we used to catch here so I don't want to change this too much.